### PR TITLE
James/new enguage api

### PR DESCRIPTION
**Pull request by JamesDHW**

### CODE REVIEW

File: opt/api/utils/http/HttpStringBuilder.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
+package opt.api.utils.http;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+
+public class HttpStringBuilder {
```
Comment (JamesDHW) [inline opt/api/utils/http/HttpStringBuilder.java:6]: This is the file I'm stuck with

File: opt/api/utils/http/HttpStringBuilder.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    
+    public static String build(BufferedReader in) {
+        StringBuilder requestBuilder = new StringBuilder();
+        String line;
+
+        try {
+            while ((line = in.readLine()) != null && !line.trim().isEmpty()) {
```
Comment (JamesDHW) [inline opt/api/utils/http/HttpStringBuilder.java]: The entire request looks like:
```
POST HTTP/1.1 /route
Header-One: Value
Header-Two: Value

{"body": "value"}
```

There's a blank line above the body (which is what I want to read) but `in.readLine()` takes in the next line, which is blank and then sits waiting for input above the body. There doesn't seem to be a way of making it move onto the next line and reading all the body

File: opt/api/utils/RequestHandler.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+             BufferedReader   in  = new BufferedReader( new InputStreamReader(connection.getInputStream()) );
+			 DataOutputStream out = new DataOutputStream(connection.getOutputStream());
+            ) 
+		{
+			String request = HttpStringBuilder.build(in);
+
+			audit.debug( String.format("Request is: %s\n", request) );
+			Audit.log( String.format("processRequest: request=%s\n", request) );
+			
+			Map<String, String> meta = HttpParser.parseMeta(request);
```
Comment (JamesDHW) [inline opt/api/utils/RequestHandler.java]: `meta` should be called `headers` everywhere

File: opt/api/actions/ActionHandler.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
+package opt.api.actions;
+
+import java.util.Map;
+
+import opt.api.utils.http.HttpException;
+
+
+public abstract class ActionHandler {
```
Comment (JamesDHW) [inline opt/api/actions/ActionHandler.java:8]: I'm pretty happy with this + `Interpret` + `Router` - makes it super easy to add new routes and `GET`/`POST` actions super easily by just adding a method or new enum option

File: opt/api/README.md [hunk 5]
```diff
@@ -0,0 +1,23 @@
+## WebServer
+
+## Run Locally
+
+### With Docker
+
+1. Build the ./DockerFile: `docker build ../.. -t enguage -f Dockerfile`
```
Comment (JamesDHW) [inline opt/api/README.md:16]: Run this within `opt/api` then you should be able to run `curl -X POST -H "Content-Type: application/json" -d '{"sessionId": "1", "utterance": "what do I need?"}' http://localhost:8080/interpret`

Comment (martinwheatman) [reply opt/api/README.md:16]: Yep, that's brill!

File: opt/api/actions/ActionHandler.java (hunk 4 above)
Comment (martinwheatman) [reply opt/api/actions/ActionHandler.java:8]: Yeah, my first attempt was simply a bare-bones server (no POSTs!)

File: opt/api/utils/http/HttpStringBuilder.java (hunk 2 above)
Comment (martinwheatman) [reply opt/api/utils/http/HttpStringBuilder.java]: I've pull this and have a look!

Comment (martinwheatman) [reply opt/api/utils/http/HttpStringBuilder.java]: It doesn't hang, the while completes at the blank line.
Remove the second condition in the while: line/trim/isEmpty.
A main program to test this function:

    public static void main(String[] args) {
    	String request = 
    			"POST HTTP/1.1 /route\n"
    			+"Header-One: Value\n"
    			+"Header-Two: Value\n"
    			+"\n"
    			+"{'body': 'value'}";
    	Reader inputString = new StringReader( request );
    	BufferedReader br = new BufferedReader( inputString );
    	String output = build( br );
    	System.out.println( "request is\n"+ output );
    }

**Review by martinwheatman — COMMENTED** [review]
I'm impressed!  Will be useful in getting Docker to work again, thanks.

File: opt/api/utils/http/HttpStringBuilder.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
+package opt.api.utils.http;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.Reader;
+import java.io.StringReader;
+
+public class HttpStringBuilder {
+    
+    public static String build(BufferedReader in) throws IOException {
```
Comment (JamesDHW) [inline opt/api/utils/http/HttpStringBuilder.java:10]: Fixed it now:
```
        while ((line = in.readLine()) != null && !line.trim().isEmpty()) {
            requestBuilder.append(line).append("\r\n");
        }
        String PostHeaderCrlf = "\r\n";
        requestBuilder.append(PostHeaderCrlf);

        int contentLength = getBodyContentLength(requestBuilder.toString().trim());
        char[] buffer = new char[contentLength];
        int bytesRead = in.read(buffer, 0, contentLength);
```

Why was this so hard? It was exactly what I was saying - `readLine` waiting but the POST request sends a blank line with no return char 😢

**Pull request merged**